File: 19.py
#19. Given the head of a linked list, remove the nth node from the end of the list and return its head.

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def printList(head):
    temp = head
    while(temp):
        print("{} ".format(temp.val), end="")
        temp = temp.next

def removeNthFromEnd(head, n) -> ListNode:
    temp = head
    llist = head
    tempB = llist
    length = 0
    #find the length of the given linked list
    while(temp):
        length+=1
        temp = temp.next
    #delete the first node of the linked list
    if(length - n == 0 and tempB != None):
        llist = tempB.next
    count = 0
    #delete the nth node from the linked list
    while(tempB):
        if(count == length - n - 1):
            delNode = tempB.next
            tempB.next = delNode.next
            break
        tempB = tempB.next
        count += 1
    return llist

if __name__ == "__main__":
    '''
    testcase : head : 1 -> 2 -> 3 -> 4 -> 5 , n = 2
    result : 1 -> 2 -> 3 -> 5
    '''

    #setups
    head = ListNode(1)
    a = ListNode(2)
    b = ListNode(3)
    c = ListNode(4)
    d = ListNode(5)
    head.next = a
    a.next = b
    b.next = c
    c.next = d

    printList(removeNthFromEnd(head, 2))

File: golang/19.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func (n ListNode) PrintList() {
	for ; n.Next != nil; n = *n.Next {
		fmt.Print(n.Val, "->")
	}
	fmt.Print(n.Val)
	fmt.Println()
}

func main() {
	d := ListNode{Val: 5}
	c := ListNode{4, &d}
	b := ListNode{3, &c}
	a := ListNode{2, &b}
	head := ListNode{1, &a}
	removeNthFromEnd(&head, 2).PrintList()
	a = ListNode{Val: 2}
	head = ListNode{1, &a}
	removeNthFromEnd(&head, 1).PrintList()
	head = ListNode{Val: 1}
	removeNthFromEnd(&head, 1).PrintList() // return nil
}

// DESC: remove the nth node from the end of the linkedlist.
// explaination: first find the length of the linkedlist, next traverse to the
// node before the node that we need to remove then remove the wanted node by doing
// curr.next = curr.next.next then we return the head of the linkedlist
// Time Complex: O(n) [I think...]
func removeNthFromEnd(head *ListNode, n int) *ListNode {
	curr := head
	count := 0
	fmt.Println("length : ", findLength(head))
	length := findLength(head)
	// for the case that the wanted node are the head of the linkedlist
	if n+1 > length {
		return curr.Next
	}
	for curr != nil {
		// traverse to the node before the wanted node
		if count == length-(n+1) {
			break
		}
		curr = curr.Next
		count++
	}
	curr.Next = curr.Next.Next
	return head
}

// DESC: find the length of the given linkedlist
func findLength(head *ListNode) int {
	curr := head
	count := 0
	for curr != nil {
		curr = curr.Next
		count++
	}
	return count
}
